File: crates/sima-core/src/encode.rs
//! Canonical byte encoding for identity-bearing data.
//!
//! Every structure that is ever hashed serializes through [`Enc`] and reads
//! back through [`Dec`]; this module exists so `to_le_bytes` never scatters
//! across the codebase. The format:
//!
//! - every integer little-endian at its natural width
//! - `i64` two's-complement little-endian
//! - bytes and str framed by a `u64` little-endian length prefix (str is its
//!   UTF-8 bytes)
//! - `Hash` as its 32 raw digest bytes
//! - `Option<Hash>` as a present-flag byte of value zero or one followed by the
//!   digest when present
//! - `Option<u64>` as the same present-flag byte followed by the little-endian
//!   value
//! - `f32` as its IEEE-754 bits in a little-endian `u32`
//! - an `f32` slice as those elements written back to back with no length prefix
//!   (the count is fixed by surrounding context)

use crate::error::{Error, Result};
use crate::hash::Hash;

/// Builder writing the canonical encoding into a `Vec<u8>`.
pub struct Enc {
    buf: Vec<u8>,
}

impl Enc {
    /// Starts an empty encoding.
    pub fn new() -> Self {
        Enc { buf: Vec::new() }
    }

    /// Consumes the builder, returning the encoded bytes.
    pub fn finish(self) -> Vec<u8> {
        self.buf
    }

    /// Writes a `u8`.
    pub fn u8(&mut self, v: u8) -> &mut Self {
        self.buf.push(v);
        self
    }

    /// Writes a `u16`, little-endian.
    pub fn u16(&mut self, v: u16) -> &mut Self {
        self.buf.extend_from_slice(&v.to_le_bytes());
        self
    }

    /// Writes a `u32`, little-endian.
    pub fn u32(&mut self, v: u32) -> &mut Self {
        self.buf.extend_from_slice(&v.to_le_bytes());
        self
    }

    /// Writes a `u64`, little-endian.
    pub fn u64(&mut self, v: u64) -> &mut Self {
        self.buf.extend_from_slice(&v.to_le_bytes());
        self
    }

    /// Writes an `i64`, two's-complement little-endian.
    pub fn i64(&mut self, v: i64) -> &mut Self {
        self.buf.extend_from_slice(&v.to_le_bytes());
        self
    }

    /// Writes three `u32`s in order, each little-endian.
    pub fn u32x3(&mut self, v: [u32; 3]) -> &mut Self {
        for part in v {
            self.u32(part);
        }
        self
    }

    /// Writes an `f32` as its IEEE-754 bits in a little-endian `u32`.
    pub fn f32(&mut self, v: f32) -> &mut Self {
        self.buf.extend_from_slice(&v.to_bits().to_le_bytes());
        self
    }

    /// Writes each `f32` as its little-endian bits, contiguously, with no
    /// length prefix — the element count is fixed by surrounding context.
    pub fn f32_slice(&mut self, v: &[f32]) -> &mut Self {
        self.buf.reserve(v.len() * 4);
        for &value in v {
            self.buf.extend_from_slice(&value.to_bits().to_le_bytes());
        }
        self
    }

    /// Writes a `u64` length prefix followed by the raw bytes.
    pub fn bytes(&mut self, v: &[u8]) -> &mut Self {
        self.u64(v.len() as u64);
        self.buf.extend_from_slice(v);
        self
    }

    /// Writes the string's UTF-8 bytes with the same framing as [`Enc::bytes`].
    pub fn str(&mut self, v: &str) -> &mut Self {
        self.bytes(v.as_bytes())
    }

    /// Writes the 32 digest bytes.
    pub fn hash(&mut self, v: &Hash) -> &mut Self {
        self.buf.extend_from_slice(v.as_bytes());
        self
    }

    /// Writes a present-flag byte, then the digest when present.
    pub fn opt_hash(&mut self, v: Option<&Hash>) -> &mut Self {
        match v {
            None => self.u8(0),
            Some(h) => self.u8(1).hash(h),
        }
    }

    /// Writes a present-flag byte, then the `u64` little-endian when present.
    pub fn opt_u64(&mut self, v: Option<u64>) -> &mut Self {
        match v {
            None => self.u8(0),
            Some(n) => self.u8(1).u64(n),
        }
    }
}

impl Default for Enc {
    fn default() -> Self {
        Enc::new()
    }
}

/// Cursor reading the canonical encoding with checked bounds; every reader
/// returns [`Error::Encoding`] on truncated or malformed input, never panics.
pub struct Dec<'a> {
    input: &'a [u8],
    pos: usize,
}

impl<'a> Dec<'a> {
    /// Starts a cursor at the beginning of `input`.
    pub fn new(input: &'a [u8]) -> Self {
        Dec { input, pos: 0 }
    }

    /// Advances past `n` bytes, borrowing them; errors when fewer remain.
    fn take(&mut self, n: usize) -> Result<&'a [u8]> {
        let remaining = self.input.len() - self.pos;
        if n > remaining {
            return Err(Error::Encoding(format!(
                "truncated input: need {n} bytes at offset {}, {remaining} remaining",
                self.pos
            )));
        }
        let slice = &self.input[self.pos..self.pos + n];
        self.pos += n;
        Ok(slice)
    }

    /// Reads exactly `N` bytes into an array.
    fn array<const N: usize>(&mut self) -> Result<[u8; N]> {
        let mut out = [0u8; N];
        out.copy_from_slice(self.take(N)?);
        Ok(out)
    }

    /// Reads a `u8`.
    pub fn u8(&mut self) -> Result<u8> {
        Ok(self.array::<1>()?[0])
    }

    /// Reads a little-endian `u16`.
    pub fn u16(&mut self) -> Result<u16> {
        Ok(u16::from_le_bytes(self.array()?))
    }

    /// Reads a little-endian `u32`.
    pub fn u32(&mut self) -> Result<u32> {
        Ok(u32::from_le_bytes(self.array()?))
    }

    /// Reads a little-endian `u64`.
    pub fn u64(&mut self) -> Result<u64> {
        Ok(u64::from_le_bytes(self.array()?))
    }

    /// Reads a two's-complement little-endian `i64`.
    pub fn i64(&mut self) -> Result<i64> {
        Ok(i64::from_le_bytes(self.array()?))
    }

    /// Reads three little-endian `u32`s.
    pub fn u32x3(&mut self) -> Result<[u32; 3]> {
        Ok([self.u32()?, self.u32()?, self.u32()?])
    }

    /// Reads an `f32` from its little-endian IEEE-754 bits.
    pub fn f32(&mut self) -> Result<f32> {
        Ok(f32::from_bits(u32::from_le_bytes(self.array()?)))
    }

    /// Reads `count` `f32` elements written back to back with no length
    /// prefix. The byte span is validated against the remaining input before
    /// any allocation, so an absurd `count` errors without allocating.
    pub fn f32_vec(&mut self, count: usize) -> Result<Vec<f32>> {
        let byte_len = count
            .checked_mul(4)
            .ok_or_else(|| Error::Encoding(format!("f32 count {count} overflows a byte length")))?;
        let raw = self.take(byte_len)?;
        let mut out = Vec::with_capacity(count);
        for chunk in raw.chunks_exact(4) {
            out.push(f32::from_bits(u32::from_le_bytes([
                chunk[0], chunk[1], chunk[2], chunk[3],
            ])));
        }
        Ok(out)
    }

    /// Reads a `u64` length prefix, then borrows that many bytes. The length
    /// is validated against the remaining input before any use, so an absurd
    /// prefix errors without allocating.
    pub fn bytes(&mut self) -> Result<&'a [u8]> {
        let len = self.u64()?;
        let len = usize::try_from(len)
            .map_err(|_| Error::Encoding(format!("length prefix {len} exceeds address space")))?;
        self.take(len)
    }

    /// Reads [`Dec::bytes`] framing and validates the payload as UTF-8.
    pub fn str(&mut self) -> Result<&'a str> {
        std::str::from_utf8(self.bytes()?)
            .map_err(|e| Error::Encoding(format!("string payload is not UTF-8: {e}")))
    }

    /// Reads 32 digest bytes.
    pub fn hash(&mut self) -> Result<Hash> {
        Ok(Hash::from_bytes(self.array()?))
    }

    /// Reads a present-flag byte (0 or 1), then the digest when present.
    pub fn opt_hash(&mut self) -> Result<Option<Hash>> {
        match self.u8()? {
            0 => Ok(None),
            1 => Ok(Some(self.hash()?)),
            flag => Err(Error::Encoding(format!(
                "invalid Option<Hash> flag byte {flag}, expected 0 or 1"
            ))),
        }
    }

    /// Reads a present-flag byte (0 or 1), then the little-endian `u64` when
    /// present.
    pub fn opt_u64(&mut self) -> Result<Option<u64>> {
        match self.u8()? {
            0 => Ok(None),
            1 => Ok(Some(self.u64()?)),
            flag => Err(Error::Encoding(format!(
                "invalid Option<u64> flag byte {flag}, expected 0 or 1"
            ))),
        }
    }

    /// Ends decoding, rejecting trailing bytes.
    pub fn finish(self) -> Result<()> {
        let trailing = self.input.len() - self.pos;
        if trailing != 0 {
            return Err(Error::Encoding(format!(
                "{trailing} trailing bytes after decode at offset {}",
                self.pos
            )));
        }
        Ok(())
    }
}

/// A type with a canonical byte encoding: its fields written through [`Enc`] in
/// declaration order and read back through [`Dec`]. Implemented by hand per type,
/// one [`encode`](Codec::encode)/[`decode`](Codec::decode) pair reproducing the
/// type's frozen layout; [`to_bytes`](Codec::to_bytes) and
/// [`from_bytes`](Codec::from_bytes) are the standalone forms, provided here in
/// terms of that pair.
pub trait Codec: Sized {
    /// Appends the canonical form to `enc`.
    fn encode(&self, enc: &mut Enc);

    /// Reads a canonical form written by [`encode`](Codec::encode). A type that
    /// routes decode through a validating constructor surfaces an invalid value
    /// as [`Error::Validation`]; a truncated buffer is [`Error::Encoding`].
    fn decode(dec: &mut Dec<'_>) -> Result<Self>;

    /// The standalone canonical bytes.
    fn to_bytes(&self) -> Vec<u8> {
        let mut enc = Enc::new();
        self.encode(&mut enc);
        enc.finish()
    }

    /// Parses standalone canonical bytes, rejecting trailing input.
    fn from_bytes(bytes: &[u8]) -> Result<Self> {
        let mut dec = Dec::new(bytes);
        let value = Self::decode(&mut dec)?;
        dec.finish()?;
        Ok(value)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::hash::to_hex;

    fn sample_hash(fill: &str) -> Result<Hash> {
        Hash::from_hex(&fill.repeat(Hash::LEN))
    }

    /// Composite covering every helper, reused by the round-trip and
    /// byte-stability tests.
    fn composite() -> Result<Vec<u8>> {
        let h1 = sample_hash("11")?;
        let h2 = sample_hash("22")?;
        let mut enc = Enc::new();
        enc.u8(0x01)
            .u16(0x0203)
            .u32(0x0405_0607)
            .u64(0x0809_0A0B_0C0D_0E0F)
            .i64(-2)
            .u32x3([1, 2, 3])
            .bytes(&[0xAA, 0xBB])
            .str("hi")
            .hash(&h1)
            .opt_hash(None)
            .opt_hash(Some(&h2))
            .opt_u64(None)
            .opt_u64(Some(0x1011_1213_1415_1617));
        Ok(enc.finish())
    }

    #[test]
    fn round_trip_every_helper() -> Result<()> {
        let buf = composite()?;
        let mut dec = Dec::new(&buf);
        assert_eq!(dec.u8()?, 0x01);
        assert_eq!(dec.u16()?, 0x0203);
        assert_eq!(dec.u32()?, 0x0405_0607);
        assert_eq!(dec.u64()?, 0x0809_0A0B_0C0D_0E0F);
        assert_eq!(dec.i64()?, -2);
        assert_eq!(dec.u32x3()?, [1, 2, 3]);
        assert_eq!(dec.bytes()?, &[0xAA, 0xBB]);
        assert_eq!(dec.str()?, "hi");
        assert_eq!(dec.hash()?, sample_hash("11")?);
        assert_eq!(dec.opt_hash()?, None);
        assert_eq!(dec.opt_hash()?, Some(sample_hash("22")?));
        assert_eq!(dec.opt_u64()?, None);
        assert_eq!(dec.opt_u64()?, Some(0x1011_1213_1415_1617));
        dec.finish()
    }

    /// The encoding is the identity-bearing byte layout; this pins it.
    /// Expected hex is derived by hand from the format documented in the
    /// module docs, field by field, in encoding order.
    #[test]
    fn composite_encoding_is_byte_stable() -> Result<()> {
        let expected = [
            "01",                                     // u8
            "0302",                                   // u16 LE
            "07060504",                               // u32 LE
            "0f0e0d0c0b0a0908",                       // u64 LE
            "feffffffffffffff",                       // i64 -2, two's complement LE
            "010000000200000003000000",               // [u32; 3] LE each
            "0200000000000000aabb",                   // bytes: u64 len + payload
            "02000000000000006869",                   // str: u64 len + UTF-8 "hi"
            &"11".repeat(Hash::LEN),                  // hash digest
            "00",                                     // Option<Hash> absent
            &format!("01{}", "22".repeat(Hash::LEN)), // Option<Hash> present
            "00",                                     // Option<u64> absent
            "011716151413121110",                     // Option<u64> present: flag + u64 LE
        ]
        .concat();
        assert_eq!(to_hex(&composite()?), expected);
        Ok(())
    }

    #[test]
    fn f32_round_trips_representative_values() -> Result<()> {
        let values = [
            1.5f32,
            -0.0,
            0.0,
            f32::MIN,
            f32::MAX,
            f32::INFINITY,
            f32::NEG_INFINITY,
        ];
        for v in values {
            let mut enc = Enc::new();
            enc.f32(v);
            let buf = enc.finish();
            let mut dec = Dec::new(&buf);
            // Compare by bits so -0.0 and the infinities are distinguished.
            assert_eq!(dec.f32()?.to_bits(), v.to_bits());
            dec.finish()?;
        }
        Ok(())
    }

    /// `f32` is written as its IEEE-754 bits in a little-endian `u32`. `1.5f32`
    /// has bit pattern `0x3FC00000`, little-endian bytes `00 00 c0 3f`.
    #[test]
    fn f32_encoding_is_byte_stable() {
        let mut enc = Enc::new();
        enc.f32(1.5);
        assert_eq!(to_hex(&enc.finish()), "0000c03f");
    }

    #[test]
    fn f32_slice_round_trips_without_a_length_prefix() -> Result<()> {
        let values = [1.5f32, -2.0, 0.25, 1024.0];
        let mut enc = Enc::new();
        enc.f32_slice(&values);
        let buf = enc.finish();
        // No length prefix: four contiguous 4-byte elements, 16 bytes.
        assert_eq!(buf.len(), values.len() * 4);
        let mut dec = Dec::new(&buf);
        let read = dec.f32_vec(values.len())?;
        dec.finish()?;
        let read_bits: Vec<u32> = read.iter().map(|v| v.to_bits()).collect();
        let want_bits: Vec<u32> = values.iter().map(|v| v.to_bits()).collect();
        assert_eq!(read_bits, want_bits);
        Ok(())
    }

    #[test]
    fn f32_truncation_errors_never_panics() {
        assert!(matches!(Dec::new(&[0u8; 3]).f32(), Err(Error::Encoding(_))));
        // A vector read past the remaining input errors before allocating.
        assert!(matches!(
            Dec::new(&[0u8; 7]).f32_vec(2),
            Err(Error::Encoding(_))
        ));
    }

    #[test]
    fn truncated_input_errors_never_panics() {
        // Every fixed-width reader against input one byte short.
        assert!(matches!(Dec::new(&[]).u8(), Err(Error::Encoding(_))));
        assert!(matches!(Dec::new(&[0]).u16(), Err(Error::Encoding(_))));
        assert!(matches!(Dec::new(&[0; 3]).u32(), Err(Error::Encoding(_))));
        assert!(matches!(Dec::new(&[0; 7]).u64(), Err(Error::Encoding(_))));
        assert!(matches!(Dec::new(&[0; 7]).i64(), Err(Error::Encoding(_))));
        assert!(matches!(
            Dec::new(&[0; 11]).u32x3(),
            Err(Error::Encoding(_))
        ));
        assert!(matches!(Dec::new(&[0; 31]).hash(), Err(Error::Encoding(_))));
        // Present-flag says a digest follows, but input ends.
        assert!(matches!(Dec::new(&[1]).opt_hash(), Err(Error::Encoding(_))));
        // Present-flag says a u64 follows, but input ends.
        assert!(matches!(
            Dec::new(&[1; 8]).opt_u64(),
            Err(Error::Encoding(_))
        ));
        // Length prefix itself truncated.
        assert!(matches!(
            Dec::new(&[2, 0, 0]).bytes(),
            Err(Error::Encoding(_))
        ));
    }

    #[test]
    fn length_prefix_beyond_remaining_is_rejected() {
        // Claims 5 payload bytes, provides 2.
        let mut buf = 5u64.to_le_bytes().to_vec();
        buf.extend_from_slice(&[0xAA, 0xBB]);
        assert!(matches!(Dec::new(&buf).bytes(), Err(Error::Encoding(_))));
        // Absurd length must error cleanly, not attempt allocation.
        let huge = u64::MAX.to_le_bytes();
        assert!(matches!(Dec::new(&huge).bytes(), Err(Error::Encoding(_))));
    }

    #[test]
    fn str_rejects_invalid_utf8() {
        let mut enc = Enc::new();
        enc.bytes(&[0xFF, 0xFE]);
        let buf = enc.finish();
        assert!(matches!(Dec::new(&buf).str(), Err(Error::Encoding(_))));
    }

    #[test]
    fn opt_hash_rejects_invalid_flag() {
        // Flag byte 2 followed by a full digest's worth of bytes.
        let buf = [2u8; 1 + Hash::LEN];
        assert!(matches!(Dec::new(&buf).opt_hash(), Err(Error::Encoding(_))));
    }

    #[test]
    fn opt_u64_rejects_invalid_flag() {
        // Flag byte 2 followed by a full u64's worth of bytes.
        let buf = [2u8; 9];
        assert!(matches!(Dec::new(&buf).opt_u64(), Err(Error::Encoding(_))));
    }

    #[test]
    fn finish_rejects_trailing_bytes() -> Result<()> {
        let mut enc = Enc::new();
        enc.u8(7);
        let mut buf = enc.finish();
        buf.push(0);
        let mut dec = Dec::new(&buf);
        assert_eq!(dec.u8()?, 7);
        assert!(matches!(dec.finish(), Err(Error::Encoding(_))));
        Ok(())
    }
}

File: python/sima/encode.py
"""The canonical byte encoding: :class:`Enc` writes it, :class:`Dec` reads it.

The same encoding carries identity-bearing values inside a frame and inside a
stored object, so a value encoded once means one thing everywhere:

- every integer little-endian at its natural width, ``i64`` two's-complement
- ``f32`` and ``f64`` as their IEEE-754 bits in a little-endian ``u32``/``u64``
- ``bytes`` and ``str`` framed by a ``u64`` little-endian length prefix, a
  string over its UTF-8 bytes
- a digest as its 32 raw bytes
- an optional value as a present-flag byte of 0 or 1, then the value when
  present
"""

from __future__ import annotations

import struct

__all__ = ["Dec", "Enc", "EncodingError", "HASH_LEN"]

#: Length of a digest on the wire.
HASH_LEN = 32

_U16 = struct.Struct("<H")
_U32 = struct.Struct("<I")
_U64 = struct.Struct("<Q")
_I64 = struct.Struct("<q")
_F32 = struct.Struct("<f")
_F64 = struct.Struct("<d")


class EncodingError(ValueError):
    """A payload that does not decode: truncated, malformed, or overlong."""


class Enc:
    """Builder appending the canonical encoding into a byte buffer.

    Every writer returns the builder, so a message is one chained expression;
    :meth:`finish` (or ``bytes(enc)``) takes the bytes out.
    """

    def __init__(self) -> None:
        self._buf = bytearray()

    def finish(self) -> bytes:
        """The encoded bytes."""
        return bytes(self._buf)

    def __bytes__(self) -> bytes:
        return self.finish()

    def u8(self, value: int) -> Enc:
        """Writes a ``u8``."""
        self._buf.append(_checked(value, 0, 0xFF, "u8"))
        return self

    def u16(self, value: int) -> Enc:
        """Writes a ``u16``, little-endian."""
        self._buf += _U16.pack(_checked(value, 0, 0xFFFF, "u16"))
        return self

    def u32(self, value: int) -> Enc:
        """Writes a ``u32``, little-endian."""
        self._buf += _U32.pack(_checked(value, 0, 0xFFFF_FFFF, "u32"))
        return self

    def u64(self, value: int) -> Enc:
        """Writes a ``u64``, little-endian."""
        self._buf += _U64.pack(_checked(value, 0, 0xFFFF_FFFF_FFFF_FFFF, "u64"))
        return self

    def i64(self, value: int) -> Enc:
        """Writes an ``i64``, two's-complement little-endian."""
        self._buf += _I64.pack(_checked(value, -(2**63), 2**63 - 1, "i64"))
        return self

    def f32(self, value: float) -> Enc:
        """Writes an ``f32`` as its IEEE-754 bits in a little-endian ``u32``."""
        self._buf += _F32.pack(value)
        return self

    def f64(self, value: float) -> Enc:
        """Writes an ``f64`` as its IEEE-754 bits in a little-endian ``u64``."""
        self._buf += _F64.pack(value)
        return self

    def bytes(self, value: bytes) -> Enc:
        """Writes a ``u64`` length prefix followed by the raw bytes."""
        self.u64(len(value))
        self._buf += value
        return self

    def str(self, value: str) -> Enc:
        """Writes the string's UTF-8 bytes with the framing of :meth:`bytes`."""
        return self.bytes(value.encode("utf-8"))

    def hash(self, value: bytes) -> Enc:
        """Writes the 32 digest bytes."""
        if len(value) != HASH_LEN:
            raise EncodingError(f"a digest is {HASH_LEN} bytes, got {len(value)}")
        self._buf += value
        return self

    def opt_bytes(self, value: bytes | None) -> Enc:
        """Writes a present-flag byte, then the framed bytes when present."""
        if value is None:
            return self.u8(0)
        return self.u8(1).bytes(value)

    def opt_hash(self, value: bytes | None) -> Enc:
        """Writes a present-flag byte, then the digest when present."""
        if value is None:
            return self.u8(0)
        return self.u8(1).hash(value)

    def opt_u64(self, value: int | None) -> Enc:
        """Writes a present-flag byte, then the ``u64`` when present."""
        if value is None:
            return self.u8(0)
        return self.u8(1).u64(value)


class Dec:
    """Cursor reading the canonical encoding with checked bounds.

    Every reader raises :class:`EncodingError` on truncated or malformed input;
    :meth:`finish` rejects trailing bytes, so a payload must be consumed
    exactly.
    """

    def __init__(self, data: bytes) -> None:
        self._data = data
        self._pos = 0

    def _take(self, count: int) -> bytes:
        """Advances past ``count`` bytes, returning them."""
        remaining = len(self._data) - self._pos
        if count > remaining:
            raise EncodingError(
                f"truncated input: need {count} bytes at offset {self._pos}, "
                f"{remaining} remaining"
            )
        taken = self._data[self._pos : self._pos + count]
        self._pos += count
        return taken

    def u8(self) -> int:
        """Reads a ``u8``."""
        return self._take(1)[0]

    def u16(self) -> int:
        """Reads a little-endian ``u16``."""
        return _U16.unpack(self._take(_U16.size))[0]

    def u32(self) -> int:
        """Reads a little-endian ``u32``."""
        return _U32.unpack(self._take(_U32.size))[0]

    def u64(self) -> int:
        """Reads a little-endian ``u64``."""
        return _U64.unpack(self._take(_U64.size))[0]

    def i64(self) -> int:
        """Reads a two's-complement little-endian ``i64``."""
        return _I64.unpack(self._take(_I64.size))[0]

    def f32(self) -> float:
        """Reads an ``f32`` from its little-endian IEEE-754 bits."""
        return _F32.unpack(self._take(_F32.size))[0]

    def f64(self) -> float:
        """Reads an ``f64`` from its little-endian IEEE-754 bits."""
        return _F64.unpack(self._take(_F64.size))[0]

    def bytes(self) -> bytes:
        """Reads a ``u64`` length prefix, then that many bytes."""
        return self._take(self.u64())

    def str(self) -> str:
        """Reads :meth:`bytes` framing and decodes it as UTF-8."""
        raw = self.bytes()
        try:
            return raw.decode("utf-8")
        except UnicodeDecodeError as e:
            raise EncodingError(f"string payload is not UTF-8: {e}") from e

    def hash(self) -> bytes:
        """Reads 32 digest bytes."""
        return self._take(HASH_LEN)

    def opt_bytes(self) -> bytes | None:
        """Reads a present-flag byte, then the framed bytes when present."""
        return self.bytes() if self._flag("optional-bytes") else None

    def opt_hash(self) -> bytes | None:
        """Reads a present-flag byte, then the digest when present."""
        return self.hash() if self._flag("optional-hash") else None

    def opt_u64(self) -> int | None:
        """Reads a present-flag byte, then the ``u64`` when present."""
        return self.u64() if self._flag("optional-u64") else None

    def flag(self) -> bool:
        """Reads a boolean flag byte, rejecting values other than 0 and 1."""
        return self._flag("flag")

    def finish(self) -> None:
        """Ends decoding, rejecting trailing bytes."""
        trailing = len(self._data) - self._pos
        if trailing:
            raise EncodingError(f"{trailing} trailing bytes after decode at offset {self._pos}")

    def _flag(self, what: str) -> bool:
        """Reads a present or boolean flag byte, which is 0 or 1 and nothing else."""
        byte = self.u8()
        if byte > 1:
            raise EncodingError(f"invalid {what} flag byte {byte}, expected 0 or 1")
        return byte == 1


def _checked(value: int, low: int, high: int, width: str) -> int:
    """The integer, refused when it does not fit the width it is written at."""
    if not low <= value <= high:
        raise EncodingError(f"{value} does not fit a {width}")
    return value
